Assert Lockdown Mode flag has been set before queried

pvollan opened the pull request:
#### 28654e3c8848b9f0da4b47a5d7b6bd993aeddb7a
<pre>
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::setJSCOptions):
</pre>
----------------------------------------------------------------------
#### 636b7439525b734d003a09ae629297fd3cb94400
<pre>
Assert Lockdown Mode flag has been set before queried
<a href="https://bugs.webkit.org/show_bug.cgi?id=278128">https://bugs.webkit.org/show_bug.cgi?id=278128</a>
<a href="https://rdar.apple.com/133884110">rdar://133884110</a>

Reviewed by NOBODY (OOPS!).

Assert that Lockdown Mode flag has been set before queried in the WebContent process. This patch also
initializes the Lockdown Mode flag in WebProcess earlier by using the value from the XPC initialization
message, which is sent in the XPC extra initialization data. This patch also asserts that JSC options
are set before JSC::initialize() is called. Setting JSC options after this call can lead to crashes.

* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
(WebKit::XPCServiceInitializer):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::XPCServiceInitializerDelegate::getExtraInitializationData):
(WebKit::setJSCOptions):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::getLaunchOptions):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeProcess):
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::isLockdownModeEnabled const):
</pre>

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -404,7 +404,7 @@ class WebProcess : public AuxiliaryProcess
 #endif
 
-    bool isLockdownModeEnabled() const { return m_isLockdownModeEnabled; }
+    bool isLockdownModeEnabled() const { RELEASE_ASSERT(m_isLockdownModeEnabled.has_value()); return m_isLockdownModeEnabled.value(); }
```
This RELEASE_ASSERT is redundant. std::optional already does this check for you.

pvollan replied:
Ah, good point, will fix.

Thanks for reviewing!

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -404,7 +404,7 @@ class WebProcess : public AuxiliaryProcess
     SpeechRecognitionRealtimeMediaSourceManager& ensureSpeechRecognitionRealtimeMediaSourceManager();
 #endif
 
-    bool isLockdownModeEnabled() const { return m_isLockdownModeEnabled; }
```
Don't we need to fix where `m_isLockdownModeEnabled` is initialized in this PR?

Why is this a separate PR from the other one?

pvollan replied:
Yes, good point, a fix for this is coming shortly in this PR. Only one PR should be required.

cdumez commented on Source/WebKit/Shared/AuxiliaryProcess.h:
```diff
@@ -210,6 +210,7 @@ struct AuxiliaryProcessInitializationParameters {
 #if PLATFORM(COCOA)
     SDKAlignedBehaviors clientSDKAlignedBehaviors;
 #endif
+    bool lockdownModeEnabled { false };
```
WebKit boolean variables need a prefix, per coding style.

cdumez replied:
Why do we need this on Auxiliary process? Isn't lockdown mode a thing only for the WebProcess only at the moment?

pvollan replied:
That is a good point. The other process types does not need this flag. I will update the PR.

Thanks for reviewing!

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
```diff
@@ -70,6 +70,8 @@ class XPCServiceInitializerDelegate {
     virtual bool getClientSDKAlignedBehaviors(SDKAlignedBehaviors&);
     virtual bool getExtraInitializationData(HashMap<String, String>& extraInitializationData);
 
+    bool lockdownModeEnabled() const;
```
Needs a prefix.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -360,6 +360,8 @@ WebProcess::~WebProcess()
 
 void WebProcess::initializeProcess(const AuxiliaryProcessInitializationParameters& parameters)
 {
+    m_isLockdownModeEnabled = parameters.lockdownModeEnabled;
```
This part of the PR I get. The rest of the refactoring, not so much.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -419,6 +419,17 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
+#if PLATFORM(IOS_FAMILY)
+        prewarmLogs();
+#endif
+        registerLogHook();
```
How concerned are we about logging that may occur after process launch but *before* receiving this initializeWebProcess IPC?

cdumez replied:
This looks a bit risky to me.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -419,6 +419,17 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
     // Reply immediately so that the identity is available as soon as possible.
     completionHandler(ProcessIdentity { ProcessIdentity::CurrentProcess });
 
+    m_isLockdownModeEnabled = parameters.isLockdownModeEnabled;
```
We're getting this flag from 2 places, once early on via XPC and then once again via this WebProcessCreationParameters struct. Why do we need it in 2 places?

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
```diff
@@ -88,48 +88,11 @@ void initializeAuxiliaryProcess(AuxiliaryProcessInitializationParameters&& param
 void setOSTransaction(OSObjectPtr<os_transaction_t>&&);
 #endif
 
+void setJSCOptions(bool configureJSCForTesting, bool enableLockdownMode, bool disableJIT, bool enableSharedArrayBuffer, bool disableJITCage);
```
No, it's not a good design to have a function that takes many booleans like this. It's too easy to pass parameters in the wrong order and introduce bugs.

pvollan replied:
Good point! I'll update the patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -603,8 +605,7 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
     ScriptExecutionContext::setCrossOriginMode(parameters.crossOriginMode);
-    m_isLockdownModeEnabled = parameters.isLockdownModeEnabled;
-    DeprecatedGlobalSettings::setArePDFImagesEnabled(!m_isLockdownModeEnabled);
+    DeprecatedGlobalSettings::setArePDFImagesEnabled(!m_isLockdownModeEnabled.value());
```
Can we use `isLockdownModeEnabled()` here?

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
```diff
@@ -178,6 +141,16 @@ void XPCServiceInitializer(OSObjectPtr<xpc_connection_t> connection, xpc_object_
+        bool enableSharedArrayBuffer = xpc_dictionary_get_bool(initializerMessage, "enable-shared-array-buffer");
+        bool disableJITCage = xpc_dictionary_get_bool(initializerMessage, "disable-jit-cage");
+
+        setJSCOptions(configureJSCForTesting, enableLockdownMode, disableJIT, enableSharedArrayBuffer, disableJITCage);
```
Maybe we should just pass the `initializerMessage` to `setJSCOptions()`?

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
```diff
@@ -88,48 +88,13 @@ void initializeAuxiliaryProcess(AuxiliaryProcessInitializationParameters&& param
 void setOSTransaction(OSObjectPtr<os_transaction_t>&&);
 #endif
 
+enum class EnableLockdownMode: bool { Yes, No };
```
No should come first, then Yes. So that No == 0 and Yes == 1.

cdumez reviewed (approved):
LGTM with nit.

pvollan committed:
Assert Lockdown Mode flag has been set before queried
https://bugs.webkit.org/show_bug.cgi?id=278128
rdar://133884110

Reviewed by Chris Dumez.

Assert that Lockdown Mode flag has been set before queried in the WebContent process. This patch also
initializes the Lockdown Mode flag in WebProcess earlier by using the value from the XPC initialization
message, which is sent in the XPC extra initialization data.

* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
(WebKit::XPCServiceInitializer):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::setJSCOptions):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::getLaunchOptions):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeProcess):
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/282363@main
